### PR TITLE
[vouchers] error moving between summary and cart pages

### LINKED ISSUE #10857: [Vouchers] Errors when/from iterating between /summary and /cart page
## Description

_This is treated as one issue, as it seems to me (I might be wrong) that this is all coming from the same root cause. Also related to #10855_

After adding a voucher and while on the confirmation page, the user may opt for returning to either the /cart page (editing quantities, for ex.):

- If there are no taxes on the order/voucher -> voucher is removed from the order
- If there are taxes on the order/voucher, and `voucher amount` > `order.total` -> snail
- If there are taxes on the order/voucher, `voucher amount` <= `order.total` -> additional parcels will be created for the voucher, on each /summary page visit, till a snail is triggered

**Bugsnag**
https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/6464d03f630d0f000895ebe9?filters[event.since]=30d&filters[error.status]=open&filters[app.release_stage]=staging&event_id=6464d90700bd31520e150000

## Expected Behavior

The user should be able to iterate between `/summary` and the `/cart` pages.

## Actual Behaviour

Iterating between the `/cart` and `/summary` pages triggers a snail.

## Steps to Reproduce

For all voucher issues: per-requisites are:
- As super admin - enable the vouchers feature toggle
- Having a hub with at least one voucher created
- The hub should have an open shop

An order/line item/customer with no taxes:

1. As a customer, on the shopfront, add items to the cart and proceed to the `/summary` step

image

2. Open the side menu and visit the /cart page. 

image

**no taxes on the order/voucher** 

3. Click Checkout button, to return to the /summary page -> notice the voucher was removed from the order.

image

**taxes on the order/voucher, and `voucher amount` > `order.total` -> snail**

3. Open the side menu and visit the /cart page. Click Checkout button -> :snail: 

image

## Animated Gif/Screenshot

image

## Workaround

- returning to the /payment step, and re-adding the voucher
- If error 500 is triggered, then the user is "stuck" -> only way out is emptying the cart on the `/cart` page

### LINKED ISSUE #11233: [Vouchers] Checkout blocked after removing items from cart (order total < 0)
## Description

The order total of a given order may become negative if, after having added a voucher, a line item quantity is reduced by editing the cart; if so, checkout is blocked for that cart.

## Expected Behavior

After reducing line item quantity order total should remain zero (after having added a voucher).

## Actual Behaviour

After reducing line item quantity order total may go to negative values (after having added a voucher).

## Steps to Reproduce

After enabling vouchers and setting up a hub to have a shopfront and a voucher of amount 10:

As a customer:
1. Add some line items to the cart -> set this up so that order is equal or lower than 10
2. Add a voucher -> order total should zero.
3. Visit the cart. 
4. Reduce the amount of line items
5. Notice the negative order total
6. Try to complete the order

## Animated Gif/Screenshot

Before editing the cart:

image

After editing the cart:

image

Trying to checkout (user stuck on `/details` step):

image

## Workaround

Can't reach /payment so can't remove voucher.

Workarounds:
- Emptying cart, and starting over
- Adding more items to the cart so that the order total > voucher amount

## Your Environment

* Version used: v4.4.2
* Browser name and version: Firefox 115
* Operating System and version (desktop or mobile): Ubuntu 22.04

### PR DESCRIPTION
**Pull request by rioug**
#### What? Why?

- Closes #10857 
- Should also fix #11233 

Matt's work in #11003 fix some part of this issue but it doesn't cover the scenario when there is a tax included or excluded from the price.
Reviewers, new code starts here : 9fccea335063b13dc9b8e7e37c8a1cac4d70d556

#### What should we test?
Assuming you have a shop with a voucher set up, follow the scenario below for voucher with no tax, with tax included in price and tax excluded from the price: 
- Go through checkout and add a voucher at the payment step
- Proceed to summary and click on "edit order details" 
- Go back to summary step by clicking checkout button 
  -> Check the voucher is still displayed properly 

##### Additional test :
As above proceed to the summary step, and then go to the /cart page, and then follow the scenario below, ,again for voucher with no tax, with tax included in price and tax excluded from the price:
- Update quantity 
  -> check total and voucher are updated accordingly
 - Click on the checkout button 
 - Go through checkout steps again 
  -> Check the voucher is still there on the payment step
  -> Check the voucher is correct on the summary step

##### Payment fees and voucher
Payment fees are calculated based on the item total, and are recalculated when confirming the order. Vouchers are calculated based on the order total. The main thing to check here is to make sure 
* when a voucher cover the order total or more, payment not is required, so the payment fee shouldn't apply
* when using a percentage based payment fee, the payment fee shouldn't change when confirming an order with a voucher

#### Release notes

Changelog Category: User facing changes 

The title of the pull request will be included in the release notes.

#### Dependencies
This PR ~~#11003~~ #11135 need to be merged first

### CODE REVIEW

File: app/controllers/spree/orders_controller.rb [hunk 1]
```diff
@@ -69,6 +69,10 @@ def update
       if @order.contents.update_cart(order_params)
         @order.recreate_all_fees! # Enterprise fees on line items and on the order itself
 
+        # Re apply the voucher
+        VoucherAdjustmentsService.calculate(@order)
+        @order.update_totals_and_states
```
Comment (dacook) [inline app/controllers/spree/orders_controller.rb:74]: > This is tested with system test

For the next reviewer: the system test is in the next commit

File: spec/system/consumer/split_checkout_tax_incl_spec.rb [hunk 2]
```diff
@@ -139,11 +139,65 @@
...
+            # Click on edit link
+            within "div", text: /Order details/ do
+              # It's a bit brittle, but the scoping doesn't seem to work
+              all(".summary-edit").last.click
+            end
+
+            # Update quantity
+            within ".cart-item-quantity" do
+              input = find(".line_item_quantity")
+              input.send_keys :up
```
Comment (dacook) [inline spec/system/consumer/split_checkout_tax_incl_spec.rb:174]: I wonder the method used here works better? 

https://github.com/openfoodfoundation/openfoodnetwork/blob/master/spec/system/consumer/shopping/cart_spec.rb#L204-L206

**Review by dacook — CHANGES_REQUESTED** [review]
Looks like a good fix! I haven't reviewed the dependent PR though because it didn't seem ready.

It would be good to fix the typo, and I've left some comments to consider for bonus points :)

File: app/services/voucher_adjustments_service.rb [hunk 3]
```diff
@@ -23,33 +23,35 @@ def self.calculate(order)
       handle_tax_included_in_price(order, amount)
     else
       adjustment.amount = amount
+      adjustment.save
```
Comment (mkllnk) [inline app/services/voucher_adjustments_service.rb:26]: I wouldn't expect `save` in a `calculate` method.

File: app/services/voucher_adjustments_service.rb [hunk 4]
```diff
@@ -23,33 +23,35 @@ def self.calculate(order)
...
-    order.adjustments.create(adjustment_attributes)
+
+    # Update the amount if tax adjustment already exist, create if not
+    tax_adjustment = order.adjustments.find_or_initialize_by(adjustment_attributes)
+    tax_adjustment.amount = tax_amount
+    tax_adjustment.save
 
     # Update the adjustment amount
-    amount = voucher_rate * (order.total - order.additional_tax_total)
+    amount = voucher_rate * (order.pre_discount_total - order.additional_tax_total)
```
Comment (mkllnk) [inline app/services/voucher_adjustments_service.rb:54]: You are re-assigning a variable given as a parameter here. I'm confused. I think the first is the voucher amount (already clamped to the order amount) and the second is the adjustment amount.

Just going through it in my head: If the voucher is more than the total there's additional tax of 10% then the voucher value will still be clamped to the pre-tax total. The voucher rate will be 100% which will create a tax adjustment to offset the tax completely. The order total will come down to zero. :heavy_check_mark:

File: spec/services/voucher_adjustments_service_spec.rb [hunk 5]
```diff
@@ -53,18 +51,14 @@
...
+        # -10 / 160 = -0.0625
         # included_tax = voucher_rate * order.included_tax_total
-        # -0.66666666 * 10 = -0.67
-        expect(subject.included_tax.to_f).to eq(-0.67)
-      end
-
-      it 'moves the adjustment state to closed' do
-        expect(subject.state).to eq('closed')
+        # -0.625 * 10 = -0.63
+        expect(subject.included_tax.to_f).to eq(-0.63)
```
Comment (mkllnk) [inline spec/services/voucher_adjustments_service_spec.rb:57]: :question: 
May brain computes this:
```
-0.625 * 10 = -6.25
```

The order total is 110, the included tax is 10. The voucher rate is 0.0625, ah, so it should be:
```suggestion
        # -0.0625 * 10 = -0.625
        expect(subject.included_tax.to_f).to eq(-0.63)
```

The rounding can then be observed.

File: app/controllers/spree/orders_controller.rb (hunk 1 above)
Comment (mkllnk) [inline app/controllers/spree/orders_controller.rb:74]: Shouldn't this happen after updating the payment fees? Ah, no, this can reduce the payment. Argh, so, do we exclude payment fees from the voucher calculation?

File: spec/system/consumer/split_checkout_tax_incl_spec.rb [hunk 6]
```diff
@@ -137,11 +142,65 @@
...
+          let!(:voucher) do
+            create(:voucher, code: 'good_code', enterprise: distributor, amount: 15)
+          end
+
+          it "recalculate the tax component properly" do
+            visit checkout_step_path(:details)
+            proceed_to_payment
+
+            # add Voucher
+            fill_in "Enter voucher code", with: voucher.code
```
Comment (mkllnk) [inline spec/system/consumer/split_checkout_tax_incl_spec.rb:158]: I would spell it out.
```suggestion
            fill_in "Enter voucher code", with: "good_code"
```

**Review by mkllnk — COMMENTED** [review]
I was about to approve but changes were requested already. So I'll move this back to In Dev.

File: app/services/voucher_adjustments_service.rb (hunk 3 above)
Comment (rioug) [reply app/services/voucher_adjustments_service.rb:26]: What would you do ? rename the method ? we need to save the result of the calculation at some point

File: spec/system/consumer/split_checkout_tax_incl_spec.rb (hunk 2 above)
Comment (rioug) [reply spec/system/consumer/split_checkout_tax_incl_spec.rb:174]: I changed it to the other method as it's more explicit, but in the end I think it's probably same same.

File: app/services/voucher_adjustments_service.rb (hunk 4 above)
Comment (rioug) [reply app/services/voucher_adjustments_service.rb:54]: Indeed it's a bit confusing, I renamed it to `adjustment_amount`

File: spec/services/voucher_adjustments_service_spec.rb [hunk 7]
```diff
@@ -87,28 +83,34 @@
...
-        expect(tax_adjustment.amount.to_f).to eq(-0.68)
+        expect(tax_adjustment.amount.to_f).to eq(-0.64)
         expect(tax_adjustment.label).to match("Tax")
       end
 
-      it 'moves the adjustment state to closed' do
-        adjustment = order.voucher_adjustments.first
-        expect(adjustment.state).to eq('closed')
+      context "when re calculating" do
+        it "update the tax asjustment" do
```
[the remark this answers is not in the document]
Comment (rioug) [reply spec/services/voucher_adjustments_service_spec.rb]: Done !

File: app/models/spree/order/checkout.rb [hunk 8]
```diff
@@ -81,6 +81,12 @@ def self.define_state_machine!
                 order.create_tax_charge!
                 order.update_totals_and_states
               end
+
+              after_transition to: :confirmation do |order|
+                VoucherAdjustmentsService.calculate(order)
+                order.update_totals_and_states
+              end
```
[the remark this answers is not in the document]
Comment (rioug) [reply app/models/spree/order/checkout.rb:88]: This covered by system specs. Your comment reminded me I forgot to remove the calculation from `VoucherAdjustmentsController` which I have now done here : https://github.com/openfoodfoundation/openfoodnetwork/pull/11117/commits/82595e307883d5049c86e2d1050f66f556355566 .
So now if the transition is broken the voucher won't be applied properly and the system specs will pick that up.

File: app/controllers/spree/orders_controller.rb (hunk 1 above)
Comment (rioug) [reply app/controllers/spree/orders_controller.rb:74]: So if I understand correctly, 
* the payment fees will be applied when we move to the confirmation step 
* voucher calculation happen after transitioning to the confirmation step (via state machine `after_transition` hook) -> so payment fees are included in the voucher calculation

The code above is used when updating the order from the `/cart` page, and as far as I can see updating anything reset the checkout flow to "details". So we end up in the scenario above and payment fee will be included in the voucher calculation.

Comment (mkllnk) [reply app/controllers/spree/orders_controller.rb:74]: Okay, so let's play this through. I have an order for $100. The Stripe payment adds a 2% fee: $2. So the total then is $102.

Then the a 10% voucher is applied, value would be $10.20. And the amount to pay reduces to $91.80.

I think that the checkout creates or updates a payment with its fee. That reduces the payment fee from $2 to $1.836 -> $1.84 and it updates the order total by 16 cents to $91.64. That amount should get charged. We have an unfortunate interdependence between voucher value and payment fee which can't be resolved completely.

Once an order is complete, there's a before_save callback to update payment fees and the same is happening on the cart page. It makes it more confusing but we don't really need to look at that now, i think.

In summary, I think that we are doing the best we can. Oh, wait, a new thought: we need to calculate the payment fee to be displayed to the user but the payment fee is really based on the payment which happens last. They were only introduced because of the fees by Stripe and Paypal. So if we exclude payment fees from the voucher calculation and calculate the payment fee after the voucher was applied then we should have the correct amount, right?

Comment (rioug) [reply app/controllers/spree/orders_controller.rb:74]: :exploding_head: I'll have another look

Comment (rioug) [reply app/controllers/spree/orders_controller.rb:74]: So I dig into that, to the extend of my knowledge payment fees are calculated based on total of items, excluding any other fees. So even in the case of a voucher being applied, the payment fee won't change when it's recalculated once the order is completed.
To take an example : 
With Order $100 (we exclude any other fees for the sake of simplicity), a Voucher of 10% and a Payment fee of 10%
- We apply the payment fee => 100 + (100 * 0.1) = 110
- We apply the voucher  => 110 - (110 * 0.1) = 99
=> Order break down : $100 product + $10 payment fees - $11 voucher = $99

Once the order is completed, the payment fees are recalculated, be because the voucher doesn't affect the item total it stays the same. So I think we are good ?

Comment (rioug) [reply app/controllers/spree/orders_controller.rb:74]: @lin-d-hop @kirstenalarsen could you confirm if the above is correct ?

Comment (lin-d-hop) [reply app/controllers/spree/orders_controller.rb:74]: @rioug this looks right to me. I agree with your findings of fees calculated on line items and thus not creating the circular behaviour that was feared.
The key here will be for rigourous @openfoodfoundation/testers ;)
We need to test the edge cases where the total is close to zero. We need to test updates to orders throughout the process to make sure all fees and vouchers are updated. Perhaps off the back of this @rioug you could spend a few minutes thinking of edge cases and confirming the 'what to test' section is right :pray:

Comment (rioug) [reply app/controllers/spree/orders_controller.rb:74]: OK, I updated the testing notes to include payment fees, and I also updated the testing notes for #10821 which is really the PR where this discussion should be taking place,

**Comment by dacook** [comment]
Just pointing out that this depends on #11135 which hasn't been merged yet.

**Comment by dacook** [comment]
@rioug now the dependent PR has been merged, it looks like there are some conflicts to fix.

File: app/controllers/spree/orders_controller.rb (hunk 1 above)
Comment (mkllnk) [reply app/controllers/spree/orders_controller.rb:74]: Ah, okay, I had the wrong assumption of the payment fee calculation then. I'm glad it's not an issue.

File: spec/system/consumer/split_checkout_tax_incl_spec.rb [hunk 9]
```diff
@@ -190,4 +244,11 @@ def assert_db_no_tax_incl
     expect(order_outside_zone.included_tax_total).to eq(0.0)
     expect(order_outside_zone.additional_tax_total).to eq(0.0)
   end
+
+  def assert_db_voucher_adjustment(amount, tax_amount)
+    adjustment = order_within_zone.voucher_adjustments.first
+    #adjustment.reload
```
Comment (mkllnk) [inline spec/system/consumer/split_checkout_tax_incl_spec.rb]: This line is removed in another commit.

**Review by mkllnk — APPROVED** [review]
Great work!

File: app/services/voucher_adjustments_service.rb (hunk 3 above)
Comment (dacook) [reply app/services/voucher_adjustments_service.rb:26]: > I would probably call it `update`.

I think that's worth doing ☝️

File: app/services/voucher_adjustments_service.rb [hunk 10]
```diff
@@ -23,42 +23,44 @@ def self.calculate(order)
...
+      state: 'open',
       tax_category: nil,
       included_tax: 0
     }
-    order.adjustments.create(adjustment_attributes)
+
+    # Update the amount if tax adjustment already exist, create if not
+    tax_adjustment = order.adjustments.find_or_initialize_by(adjustment_attributes)
+    tax_adjustment.amount = tax_amount
+    tax_adjustment.save
```
Comment (dacook) [inline app/services/voucher_adjustments_service.rb:51]: There's a couple of things happening in this method, and maybe it could be clearer if they were split into two methods.

I think the tax adjustment could be moved to a private method, something like:
```suggestion
    update_tax_adjustment_for(adjustment, tax_amount)
```

**Review by dacook — APPROVED** [review]
I think this is good, but can see potential to arrange the code, do you think these will be easy to update?

Otherwise feel free to move to testing.

**Review by dacook — APPROVED** [review]
Good one! 👍

**Comment by drummer83** [comment]
## Testing notes BEFORE staging
Testing AFTER staging will follow.

| Test case | Tested for 'No tax', 'Tax incl. in price', 'Tax excl. from price' |
| - | - |
| Moving between summary and VIEWING cart | :heavy_check_mark: |
| Moving between summary and INCREASING cart | :x: Voucher value remains fixed to the value before editing the order |
| Moving between summary and REDUCING cart | :x: Redirection to home but placing the order (vid 20-40) without user confirmation, incorrect confirmation email with credit owed, incorrect payment listed in confirmation email |
| Voucher removed from order when moving between summary and cart? | No. :heavy_check_mark: |
| Order total < voucher amount | :x: Payment fee listed in order details, not shown in summary, but subtracted from voucher amount (img 1); Negative order total if payment method had to be selected due to General Issue 1 |
| Order total > voucher amount | :heavy_check_mark: |
| Order total < 0 due to editing order (#11233) | :x: Redirection to home but placing the order (vid 20-40) without user confirmation, incorrect confirmation email with credit owed, incorrect payment listed in confirmation email |
| Payment fee doesn't change when applying the voucher | :heavy_check_mark: |
| Payment fee is not applied when order total <= voucher amount | :question: No. Fee is still applied if a payment method was selected earlier. (img 2) - Couldn't reproduce this myself. :cry: |

Img 1: 
image

Img 2: 
image


### General issues BEFORE staging
- Going back to Step 2 after being on step 3 already shows 'no payment required', but then an error is displayed because no payment method is selected.
- Editing an order on /cart page doesn't update transaction fees. This happens only after proceeding to checkout again.
- 'No payment required' remains visible after removing a voucher code. Error message when clicking Next.

**Comment by drummer83** [comment]
Hi @rioug,
Thanks for continuing the work on vouchers! :muscle: 
Here are my notes after staging this PR.

## Testing notes AFTER staging

| Test case | Tested for 'No tax', 'Tax incl. in price', 'Tax excl. from price' |
| - | - |
| Moving between summary and VIEWING cart | :heavy_check_mark: |
| Moving between summary and INCREASING cart | :x: 1st checkout: even if order total < voucher amount a payment method is required (should be 'no payment required'), and it's listed on step 3 (img 4); After increasing the cart (still < voucher) the 'no payment required' is displayed, but then error message and payment method needs to be selected, then on step 3 a payment method and transaction fee is displayed even though the voucher covers the order total; After confirmation incorrect order total (transaction fee is balance due) (vid 10-49) |
| Moving between summary and REDUCING cart | :x: 1st checkout: even if order total < voucher amount a payment method is required (should be 'no payment required'), and it's listed on step 3; After reducing the cart (still < voucher) the 'no payment required' is displayed, but then error message and payment method needs to be selected, then on step 3 a payment method and transaction fee is displayed even though the voucher covers the order total; After confirmation incorrect order total (transaction fee is balance due) (vid 11-05) |
| Voucher removed from order when moving between summary and cart? | No. :heavy_check_mark: |
| Order total < voucher amount | :x: Payment fee listed in order details, and shown in summary (img 3); order total = 0 (this is ok), but payment fee is listed as not paid after confirming the order; |
| Order total > voucher amount | :heavy_check_mark: |
| Order total < 0 due to editing order (#11233) | :heavy_check_mark: |
| Payment fee doesn't change when applying the voucher | :heavy_check_mark: |
| Payment fee is not applied when order total <= voucher amount | :x: Payment fee is still applied (img 4). |

**Img 3:**
image

**Img 4:**
image

**Vid 10-49**

https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/0fdacb3a-c9c4-41ef-aaf2-45e03a169bb3

**Vid 11-05**

https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/5b5db1e7-4153-40b2-84a8-c8c2f8aa574d


### General issues AFTER staging
- Going back to Step 2 after being on step 3 already shows 'no payment required', but then an error is displayed because no payment method is selected. https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/dc9e814a-658d-4591-ae77-47eb1a77823b
- Editing an order on /cart page doesn't update transaction fees. This happens only after proceeding to checkout again. https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/811c0ac3-9fe3-4e4e-a39a-37384f441038
- 'No payment required' remains visible after removing a voucher code. Error message when clicking Next.
- Adding the voucher on step 2 after step 3 had been visited already ignores the voucher. https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/b44bf5bf-7728-462a-92c9-484f80d3fc85
- Editing the cart after completing step 1 and 2 of checkout already, redirects to step 1 when clicking checkout. This has already been completed (as step 2 has been as well), so shouldn't we redirect to step 3 instead? Shipping method should be remembered.
- Tax calculation for tax excluded from price changed: 
  - BEFORE: Tax on voucher was calculated **_excluding_** the transaction fee.
image
  - AFTER: Tax on voucher is calculated **_including_** the transaction fee.
image

# Summary
- It looks like this PR is solving #11233. :heavy_check_mark: 
- Tax calculation for tax excluded from price changed. Was this intentional and decided somewhere? :question: 
- There are some specific and some general issues which should be fixed.

My proposal is to merge this one and create issues for the open bits. Let me know what you think please and feel free to merge if you agree.

Thanks again!

**Comment by rioug** [comment]
Thanks for the thorough testing @drummer83 !

> Editing the cart after completing step 1 and 2 of checkout already, redirects to step 1 when clicking checkout. This has already been completed (as step 2 has been as well), so shouldn't we redirect to step 3 instead? Shipping method should be remembered.

I believe this is an existing behaviour, nothing to do with voucher. But I agree it's a little bit confusing, it should at least redirect to step 2 in case you need to change payment method

> Tax calculation for tax excluded from price changed: 
I don't think that was intentional, from memory the transaction fee was excluded from voucher calculation, but I haven't had time to double check that. 

I fixed some of the issues, but I am still looking at anything related to payment fees.

**Comment by rioug** [comment]
@filipefurtad0 I am trying to create an order with a payment and payment fee. I am not quite sure how to create the payment fee, do you have any pointers ?  
This is what I tried so far : 
```ruby

let(:bogus_payment) { create(:bogus_payment_method, distributors: [create(:enterprise)]) }

before do
  payment = create(:payment, amount: 1, payment_method: bogus_payment)
  order.payments << payment
  order.save!
  order.update_payment_fees!
end
```
I think I need to setup some sort of fees on the payment ?

**Comment by drummer83** [comment]
Ok, looks like you keep working within this PR so I will move back to In Dev.

**Comment by filipefurtad0** [comment]
Hey @rioug,

>  This is what I tried so far :

For a transaction fee to be applied, in addition to the payment method, we'd need to set up a calculator.

But maybe there's an easier way, by using the [factory for orders with fees](https://github.com/openfoodfoundation/openfoodnetwork/blob/0bcde2dbda82a73ee2539ae62c167743ead620f5/spec/factories/order_factory.rb#L235) `(:completed_order_with_fees)`.

In case you need more granularity, like setting the calculator type, or values, then the approach on the [payment_spec.rb](https://github.com/openfoodfoundation/openfoodnetwork/blob/0bcde2dbda82a73ee2539ae62c167743ead620f5/spec/models/spree/payment_spec.rb#L913) would be best, I guess.

Hope this helps.

**Comment by rioug** [comment]
@drummer83 

> Editing an order on /cart page doesn't update transaction fees. This happens only after proceeding to checkout again. 

This is an existing bug so I won't be fixing it here. But as you noticed, it will be recalculated as we proceed to checkout again so it's a "minor" issue

> Tax calculation for tax excluded from price changed. Was this intentional and decided somewhere? ❓

I looked at this again, and transaction fees are included in the tax calculation for tax excluded from price. Nothing should have changed here. So the "AFTER" is correct here.

File: app/controllers/voucher_adjustments_controller.rb [hunk 11]
```diff
@@ -47,6 +47,10 @@ def add_voucher
       return false
     end
 
+    # Clear payments and payment_fee, to not affect voucher adjustment calculation
+    @order.all_adjustments.payment_fee.destroy_all
+    @order.payments.clear
```
Comment (dacook) [inline app/controllers/voucher_adjustments_controller.rb]: Sorry but I have to ask, is this potentially clearing records of actual payments? 
I'm guessing at this point the payments (and indeed order) are in an "incomplete" state, so it's ok to delete. But perhaps we should add a scope just in case? Eg:

```
    @order.payments.incomplete.clear
```

I looked at the Payment model and expected to see a `before_destroy` for this but didn't. Maybe it should be there..
@mkllnk do you have any insights?

**Review by dacook — COMMENTED** [review]
Looks good, but pending a question about the `payments.clear` call.

File: app/controllers/voucher_adjustments_controller.rb [hunk 12]
```diff
@@ -74,4 +79,10 @@ def render_error
   def voucher_params
     params.require(:order).permit(:voucher_code)
   end
+
+  # Clear payments and payment fees, to not affect voucher adjustment calculation
+  def clear_payments
+    @order.all_adjustments.payment_fee.destroy_all
+    @order.payments.clear
```
[the remark this answers is not in the document]
Comment (rioug) [reply app/controllers/voucher_adjustments_controller.rb:86]: My understanding is that payment are processed once we confirm the order, see below: 
https://github.com/openfoodfoundation/openfoodnetwork/blob/ad37c401e5866f0ae2e432c37076885a004f2a1f/app/controllers/split_checkout_controller.rb#L115-L126

and once the order is confirmed then the state machine will finalise the order. So If I understand correctly we shouldn't have any completed payments when navigating between checkout steps. 
Also if the checkout is restarted, the the user will be going through the payment step again and will be able to (re) choose the payment method. 

I added this to fix a specific scenario where the user goes through the checkout till the confirmation step, then goes back to the payment step and add a voucher covering the order amount. In this case we need to remove any existing payment/payment fee as now there is no payment required. 

Would we have an already completed payment if we face a stock conflict and have to restart the checkout ? You would think that we would check the stock situation before taking a payment. 
From what I can see, we would be redirected to cart if there is a stock issue when navigation checkout steps:
https://github.com/openfoodfoundation/openfoodnetwork/blob/ad37c401e5866f0ae2e432c37076885a004f2a1f/app/controllers/concerns/checkout_callbacks.rb#L34
https://github.com/openfoodfoundation/openfoodnetwork/blob/ad37c401e5866f0ae2e432c37076885a004f2a1f/app/controllers/concerns/checkout_callbacks.rb#L66-L70
I couldn't see any other checks so maybe I am missing something here.

Anyway, I'll apply @dacook suggestion to be on the safer side.

[the remark this answers is not in the document]
Comment (mkllnk) [reply app/controllers/voucher_adjustments_controller.rb]: > Would we have an already completed payment if we face a stock conflict and have to restart the checkout ? You would think that we would check the stock situation before taking a payment.
> From what I can see, we would be redirected to cart if there is a stock issue when navigation checkout steps

Of course we check the stock before we take a payment but there's always room for race conditions. We often process several checkouts in the same second and it can happen that two people try to checkout the same last item at the same time. It's really hard to predict what exactly happens because it depends on the timing and where the error is raised.

Actually, the CurrentOrderLocker should prevent this. So maybe it's not a problem any more. I'm just scared that this race condition will come back to haunt us... :fearful:

**Review by mkllnk — APPROVED** [review]

**Review by dacook — APPROVED** [review]
👍  Good one

**Comment by drummer83** [comment]
Hi @rioug,
We're getting closer. :+1: 
Thanks for continuing your work on this!

## Conclusion
To unblock the percentage PR and to keep things simple I'd vote to merge this one and fix the open issues in separate PRs. I will create issues for those. Please go ahead and merge if you agree or let me know if you don't.

## Testing notes AFTER staging

| Test case | Tested for 'No tax', 'Tax incl. in price', 'Tax excl. from price' |
| - | - |
| Moving between summary and VIEWING cart | :heavy_check_mark: |
| Moving between summary and INCREASING cart | :heavy_check_mark: 1st checkout: if order total < voucher amount then 'no payment required' is shown and it's listed on step 3 (img 5); :x: After editing and increasing the cart (still order total < voucher amount) the 'no payment required' is **NOT** displayed, a payment method needs to be selected and then on step 3 the payment method and transaction fee is displayed even though the voucher covers the order total (img 6); |
| Moving between summary and REDUCING cart | :heavy_check_mark: 1st checkout: if order total < voucher amount then 'no payment required' is shown and it's listed on step 3 (img 5); :x: After editing and reducing the cart (still order total < voucher amount) the 'no payment required' is **NOT** displayed, a payment method needs to be selected and then on step 3 the payment method and transaction fee is displayed even though the voucher covers the order total (img 6); |
| Voucher removed from order when moving between summary and cart? | No. :heavy_check_mark: |
| Order total > voucher amount | :heavy_check_mark: |
| Order total < 0 due to editing order (#11233) | :heavy_check_mark: |
| Payment fee doesn't change when applying the voucher | :heavy_check_mark: |
| Payment fee is not applied when order total <= voucher amount | :x: Payment fee is still applied when editing the cart after being on step 3 of checkout already (img 6). |

**Image 5:**
image

**Image 6:**
image

### General issues AFTER staging
- Found the difference between these two results. It's not related to BEFORE and AFTER staging, it depends on how you finish the checkout steps. :arrow_right: separate issue
  - Finishing **all three steps in a row** and adding the voucher in step 2:
    - **CORRECT:** Tax on voucher is calculated **_including_** the transaction fee.
image
  - Leaving step 2 without adding the voucher and then **going back from step 3 to step 2** to add the voucher yields this:
    - **NOT CORRECT:** Tax on voucher is calculated **_excluding_** the transaction fee.
image
- Going back to Step 2 after being on step 3 already shows 'no payment required', but then an error is displayed because no payment method is selected. :arrow_right: fixed :heavy_check_mark: 
- Editing an order on /cart page doesn't update transaction fees and removes the shipping fees from the table. :arrow_right: separate issue 
https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/811c0ac3-9fe3-4e4e-a39a-37384f441038
- 'No payment required' remains visible after removing a voucher code. Error message when clicking Next. :arrow_right: fixed :heavy_check_mark: 
- Adding the voucher on step 2 after step 3 had been visited already ignores the voucher. :arrow_right: fixed :heavy_check_mark: 
- Editing the cart after completing step 1 and 2 of checkout already, redirects to step 1 when clicking checkout. This has already been completed (as step 2 has been as well), so shouldn't we redirect to step 3 instead? Shipping method should be remembered. :arrow_right: separate issue
- Negative order total (credit owed) after changing to a shipping method without fee. :arrow_right: separate issue
https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/c5fc9ebc-f7f7-40ac-b2bc-5d20bd73c7ea
